java/java.lsp.server/nbcode/integration/src/org/netbeans/modules/nbcode/integration/UIDefaultsIconMetadata.java

```diff
@@ -0,0 +1,213 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.netbeans.modules.nbcode.integration;
+
+import java.awt.Component;
+import java.awt.EventQueue;
+import java.awt.Graphics;
+import java.awt.Image;
+import java.awt.image.ImageObserver;
+import java.awt.image.ImageProducer;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Enumeration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.swing.Icon;
+import javax.swing.ImageIcon;
+import javax.swing.UIDefaults;
+import javax.swing.UIManager;
+import org.openide.modules.OnStart;
+import org.openide.util.ImageUtilities;
+
+/**
+ * Goes through UIDefaults and replaces Icons and Images with variants that
+ * return image's URL from the 'url' property.
+ * @author sdedic
+ */
+@OnStart
+public class UIDefaultsIconMetadata implements Runnable {
+    private static final Logger LOG = Logger.getLogger(UIDefaultsIconMetadata.class.getName());
+    
+    public static final String PROP_URI = "uri"; // NOI18N
+    public static final String PROP_URL = "url"; // NOI18N
+    public static final String PROP_ORIGINAL_ICON = "originalIcon"; // NOI18N
+    public static final String PROP_ORIGINAL_IMAGE = "originalImage"; // NOI18N
+    
+    private static final String URN_DEFAULTS_PREFIX = "urn:uidefaults:"; // NOI18N
+    
+    private static Image wrapImage(String key, Image image, Object... meta) {
+        Object o = image.getProperty(PROP_URL, null); // NOI18N
+        if (o instanceof URL) {
+            return image;
+        }
+        
+        o = image.getProperty(PROP_URI, null);
+        if (o instanceof String || o instanceof URI) {
+            return image;
+        } else {
+            
+            try {
+                Object[] a;
+                
+                if (meta == null || meta.length == 0) {
+                    a = new Object[] { PROP_URI, new URI(URN_DEFAULTS_PREFIX + key) };
+                } else {
+                    a = Arrays.copyOf(meta, meta.length + 2);
+                    a[meta.length] = PROP_URI;
+                    a[meta.length + 1] = new URI(URN_DEFAULTS_PREFIX + key);
+                }
+                return new MetadataImage(image, a);
+            } catch (URISyntaxException ex) {
+                return image;
+            }
+        }
+    }
+    
+    @Override
+    public void run() {
+        EventQueue.invokeLater(() -> decorateUIDefaults());
+    }
+    
+    public static void decorateUIDefaults() {
+        UIDefaults defs = UIManager.getDefaults();
+        
+        for (Enumeration en = defs.keys(); en.hasMoreElements(); ) {
```

Comment:
Replacing `UIManager.getDefaults()` and decorating the original items when one asks for them would also be an option. Faster on startup, probably. Converting only the icons that are needed.